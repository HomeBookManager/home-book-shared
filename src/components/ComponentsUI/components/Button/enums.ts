export enum ButtonColor {
  primary = 'primary',
  secondary = 'secondary',
  success = 'success',
  warning = 'warning',
  error = 'error',
}

export enum ButtonSize {
  small = 'small',
  medium = 'medium',
  large = 'large',
}

export enum ButtonVariant {
  text = 'text',
  contained = 'contained',
  outlined = 'outlined',
}
