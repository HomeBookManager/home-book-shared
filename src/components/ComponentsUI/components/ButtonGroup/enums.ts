export enum ButtonGroupOrientation {
  horizontal = 'horizontal',
  vertical = 'vertical',
}
