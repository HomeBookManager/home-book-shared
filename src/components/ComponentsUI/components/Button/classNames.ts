import { mapValues } from 'lodash';

// types
import { ButtonColor, ButtonSize, ButtonVariant } from './enums';

export const className = 'Button';

export const getIconModificators = (property: 'fill' | 'stroke') => ({
  ...mapValues(ButtonColor, (color) => ({
    ...mapValues(
      ButtonVariant,
      (variant) =>
        `${className}__icon--apply-${property}--${variant}--${color}`,
    ),
  })),
});

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      fullwidth: `${className}--full-width`,
      forcedHover: {
        ...mapValues(ButtonVariant, (variant) => ({
          ...mapValues(
            ButtonColor,
            (color) => `${className}--${variant}--${color}--forced-hover`,
          ),
        })),
      },
      ...mapValues(ButtonColor, (color) => `${className}--${color}`),
      ...mapValues(ButtonSize, (size) => `${className}--${size}`),
      ...mapValues(ButtonVariant, (variant) => `${className}--${variant}`),
    },
  },
  icon: {
    name: `${className}__icon`,
    modificators: {
      applyFill: `${className}__icon--apply-fill`,
      applyStroke: `${className}__icon--apply-stroke`,
      end: `${className}__icon--end`,
      start: `${className}__icon--start`,
      ...mapValues(ButtonSize, (size) => `${className}__icon--${size}`),
    },
  },
};
