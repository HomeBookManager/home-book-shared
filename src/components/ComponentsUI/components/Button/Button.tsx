import cx from 'classnames';
import { ButtonHTMLAttributes, ReactNode, Ref, forwardRef } from 'react';

// components
import E2EDataAttribute from '../../../E2EDataAttributes/E2EDataAttribute';

// hooks
import { useButtonGroupContext } from './hooks/useButtonGroupContext';
import { useClickInteraction } from './hooks/useClickInteraction';
import { useIcon } from './hooks/useIcon';
import { useRippleEffect } from '../../../../hooks/useRippleEffect/useRippleEffect';

// others
import { className as classNameButton, classNames } from './classNames';

// styles
import './styles/button.scss';

// types
import { ButtonColor, ButtonSize, ButtonVariant } from './enums';
import { E2EAttribute } from '../../../E2EDataAttributes/enums';
import { TButtonIcon } from './types';

export type TButtonProps = Omit<
  ButtonHTMLAttributes<HTMLElement>,
  'disabled'
> & {
  children?: ReactNode;
  className?: string;
  color?: ButtonColor;
  disabled?: boolean;
  disabledRippleEffect?: boolean;
  e2eAttribute?: string;
  e2eValue?: number | string;
  endIcon?: TButtonIcon;
  forcedHover?: boolean;
  fullWidth?: boolean;
  onClick?: () => void;
  ref?: Ref<HTMLButtonElement>;
  size?: ButtonSize;
  startIcon?: TButtonIcon;
  variant?: ButtonVariant;
};

export const Button = forwardRef<HTMLButtonElement, TButtonProps>(
  (
    {
      children,
      className = '',
      color = ButtonColor.primary,
      disabled = false,
      disabledRippleEffect = false,
      endIcon = null,
      e2eAttribute = E2EAttribute.button,
      e2eValue = '',
      forcedHover = false,
      fullWidth = false,
      onClick,
      size = ButtonSize.medium,
      startIcon = null,
      type = 'button',
      variant = ButtonVariant.contained,
      ...restProps
    },
    ref,
  ) => {
    const {
      classNameButtonGroup,
      colorButtonGroup,
      disabledRippleEffectButtonGroup,
      forcedHoverButtonGroup,
      sizeButtonGroup,
      variantButtonGroup,
      ...restButtonGroupProps
    } = useButtonGroupContext();

    const Icon = useIcon(size);

    const { rippleEffect, triggerRippleEffect } = useRippleEffect(
      classNames[classNameButton].name,
    );

    const onClickHandler = useClickInteraction(
      disabledRippleEffect || disabledRippleEffectButtonGroup,
      onClick,
      triggerRippleEffect,
    );

    return (
      <E2EDataAttribute type={e2eAttribute} value={e2eValue}>
        <button
          className={cx(
            className,
            classNameButtonGroup,
            classNames[classNameButton].name,
            { [classNames[classNameButton].modificators.fullwidth]: fullWidth },
            {
              [classNames[classNameButton].modificators.forcedHover]:
                forcedHover || forcedHoverButtonGroup,
            },
            classNames[classNameButton].modificators[colorButtonGroup || color],
            classNames[classNameButton].modificators[sizeButtonGroup || size],
            classNames[classNameButton].modificators[
              variantButtonGroup || variant
            ],
          )}
          disabled={disabled}
          onClick={onClickHandler}
          ref={ref}
          type={type}
          {...restButtonGroupProps}
          {...restProps}
        >
          {startIcon && <Icon placement="start" src={startIcon} />}
          {children}
          {endIcon && <Icon placement="end" src={endIcon} />}
          {rippleEffect}
        </button>
      </E2EDataAttribute>
    );
  },
);

export default Button;
