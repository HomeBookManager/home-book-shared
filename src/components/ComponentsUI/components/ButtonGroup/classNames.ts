import { mapValues } from 'lodash';

// types
import { ButtonColor, ButtonVariant } from '../Button/enums';
import { ButtonGroupOrientation } from './enums';

export const className = 'ButtonGroup';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      ...mapValues(
        ButtonGroupOrientation,
        (orientation) => `${className}--${orientation}`,
      ),
    },
  },
  button: {
    name: `${className}__button`,
    modificators: {
      forcedHover: `${className}__button--forced-hover`,
      ...mapValues(ButtonColor, (color) => `${className}__button--${color}`),
      ...mapValues(
        ButtonGroupOrientation,
        (orientation) => `${className}__button--${orientation}`,
      ),
      ...mapValues(
        ButtonVariant,
        (variant) => `${className}__button--${variant}`,
      ),
    },
  },
};
