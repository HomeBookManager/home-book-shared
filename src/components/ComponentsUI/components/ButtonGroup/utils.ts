import cx from 'classnames';

// others
import { classNames } from './classNames';

// types
import { ButtonColor, ButtonVariant } from '../Button/enums';
import { ButtonGroupOrientation } from './enums';

export const composeButtonClassNames = (
  classNameButton: string,
  color: ButtonColor,
  forcedHover: boolean,
  orientation: ButtonGroupOrientation,
  variant: ButtonVariant,
): string =>
  cx(
    classNameButton,
    classNames.button.name,
    classNames.button.modificators[color],
    {
      [classNames.button.modificators.forcedHover]: forcedHover,
    },
    classNames.button.modificators[orientation],
    classNames.button.modificators[variant],
  );
