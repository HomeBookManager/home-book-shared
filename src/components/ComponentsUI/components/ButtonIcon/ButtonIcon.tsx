import cx from 'classnames';
import { ButtonHTMLAttributes, FC, useState } from 'react';

// components
import CirclePulse, {
  TCirclePulseProps,
} from '../../../../shared/CirclePulse/CirclePulse';
import E2EDataAttribute from '../../../E2EDataAttributes/E2EDataAttribute';
import Icon, { TIconProps } from '../Icon/Icon';

// hooks
import { useClickInteraction } from './hooks/useClickInteraction';
import { useTheme } from '../../../../hooks/useTheme/useTheme';

// others
import { className as classNameButtonIcon, classNames } from './classNames';

// styles
import './styles/button-icon.scss';

// types
import { ButtonSize } from '../Button/enums';
import { E2EAttribute } from '../../../E2EDataAttributes/enums';

export type TButtonIconProps = Omit<
  ButtonHTMLAttributes<HTMLElement>,
  'children' | 'disabled'
> & {
  className?: string;
  classNameCirclePulse?: string;
  classNameIcon?: string;
  disabled?: boolean;
  disablePulseEffect?: boolean;
  e2eAttribute?: string;
  e2eValue?: number | string;
  forcedHover?: boolean;
  iconComponent: TIconProps['iconComponent'];
  size?: ButtonSize;
};

export const ButtonIcon: FC<TButtonIconProps> = ({
  className = '',
  classNameCirclePulse = '',
  classNameIcon = '',
  disabled = false,
  disablePulseEffect = false,
  e2eAttribute = E2EAttribute.buttonIcon,
  e2eValue = '',
  forcedHover = false,
  iconComponent,
  onClick = null,
  size = ButtonSize.medium,
  type = 'button',
  ...restProps
}) => {
  const { classNamesWithTheme } = useTheme(classNames);

  const [pulseElements, setPulseElements] = useState<
    TCirclePulseProps['pulseElements']
  >([]);

  const onClickHandler = useClickInteraction(
    disablePulseEffect,
    onClick,
    pulseElements,
    setPulseElements,
  );

  return (
    <E2EDataAttribute type={e2eAttribute} value={e2eValue}>
      <button
        className={cx(
          classNamesWithTheme[classNameButtonIcon].name,
          {
            [classNamesWithTheme[classNameButtonIcon].modificators.forcedHover]:
              forcedHover,
          },
          classNamesWithTheme[classNameButtonIcon].modificators[size],
          className,
        )}
        disabled={disabled}
        onClick={onClickHandler}
        type={type}
        {...restProps}
      >
        <Icon
          className={cx(
            classNameIcon,
            classNameCirclePulse,
            classNamesWithTheme.icon.name,
            classNamesWithTheme.icon.modificators[size],
          )}
          forcedHover
          iconComponent={iconComponent}
        />
        {pulseElements.map((key) => (
          <CirclePulse
            animationDuration={1000}
            className={cx(
              classNameCirclePulse,
              classNamesWithTheme.circlePulse,
            )}
            pulseElements={pulseElements}
            setPulseElements={setPulseElements}
            key={key}
          />
        ))}
      </button>
    </E2EDataAttribute>
  );
};

export default ButtonIcon;
