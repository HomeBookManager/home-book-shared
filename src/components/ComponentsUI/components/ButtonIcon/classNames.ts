import { mapValues } from 'lodash';

// types
import { ButtonSize } from '../Button/enums';

export const className = 'ButtonIcon';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      forcedHover: `${className}--forced-hover`,
      ...mapValues(ButtonSize, (size) => `${className}--${size}`),
    },
  },
  icon: {
    name: `${className}__icon`,
    modificators: {
      ...mapValues(ButtonSize, (size) => `${className}__icon--${size}`),
    },
  },
  circlePulse: `${className}__circle-pulse`,
};
