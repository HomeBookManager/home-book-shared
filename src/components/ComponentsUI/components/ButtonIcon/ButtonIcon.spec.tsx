import { fireEvent, render, waitFor } from '@testing-library/react';

// components
import ButtonIcon from './ButtonIcon';
import IconUpOutlined from '../Icon/components/IconUpOutlined';

// others
import { className as classNameButtonIcon, classNames } from './classNames';

// types
import { ButtonSize } from '../Button/enums';
import { E2EAttribute } from '../../../E2EDataAttributes/enums';

// utils
import { enumToArray } from '../../../../utils';
import { getByE2EAttribute } from '../../../../tests';
import { getDataTestAttribute } from '../../../E2EDataAttributes/utils';

const className = 'className';
const classNameCirclePulse = 'classNameCirclePulse';
const classNameIcon = 'classNameIcon';
const iconComponent = IconUpOutlined;

describe('ButtonIcon behaviors', () => {
  it('should render pulseEffect after click', async () => {
    // before
    const { container } = render(
      <ButtonIcon
        classNameCirclePulse={classNameCirclePulse}
        iconComponent={iconComponent}
      />,
    );
    const buttonIcon = getByE2EAttribute(container, E2EAttribute.buttonIcon);

    // action
    fireEvent.click(buttonIcon);

    // result
    await waitFor(() => {
      expect(buttonIcon.lastChild).toHaveClass(classNameCirclePulse);
    });
  });
});

describe('ButtonIcon props', () => {
  const mockCallBack = jest.fn();

  it('should pass className', () => {
    // before
    const { container } = render(
      <ButtonIcon className={className} iconComponent={iconComponent} />,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.buttonIcon)).toHaveClass(
      className,
    );
  });

  it('should pass classNameCirclePulse', () => {
    // before
    const { container } = render(
      <ButtonIcon
        classNameCirclePulse={classNameCirclePulse}
        iconComponent={iconComponent}
      />,
    );

    // result
    expect(container.querySelector(`.${classNameCirclePulse}`)).toHaveClass(
      classNameCirclePulse,
    );
  });

  it('should pass classNameIcon', () => {
    // before
    const { container } = render(
      <ButtonIcon
        classNameIcon={classNameIcon}
        iconComponent={iconComponent}
      />,
    );

    // result
    expect(container.querySelector(`.${classNameIcon}`)).toHaveClass(
      classNameIcon,
    );
  });

  it('should pass disabled', () => {
    // before
    const { container } = render(
      <ButtonIcon disabled iconComponent={iconComponent} />,
    );

    // result
    expect(
      getByE2EAttribute(container, E2EAttribute.buttonIcon),
    ).toHaveAttribute('disabled');
  });

  it('should pass disablePulseEffect', () => {
    // before
    const { container } = render(
      <ButtonIcon disablePulseEffect iconComponent={iconComponent} />,
    );

    // find
    const buttonIcon = getByE2EAttribute(container, E2EAttribute.buttonIcon);

    // action
    fireEvent.click(buttonIcon);

    // result
    expect(buttonIcon.lastChild).toHaveTextContent('');
  });

  it('should pass e2eAttribute', () => {
    // before
    const { container } = render(
      <ButtonIcon
        e2eAttribute={E2EAttribute.buttonIcon}
        iconComponent={iconComponent}
      />,
    );

    // result
    expect(
      getByE2EAttribute(container, E2EAttribute.buttonIcon),
    ).toHaveAttribute(getDataTestAttribute(E2EAttribute.buttonIcon));
  });

  it('should pass e2eValue', () => {
    // before
    const { container } = render(
      <ButtonIcon e2eValue="e2eValue" iconComponent={iconComponent} />,
    );

    // result
    expect(
      getByE2EAttribute(container, E2EAttribute.buttonIcon),
    ).toHaveAttribute(
      getDataTestAttribute(E2EAttribute.buttonIcon),
      'e2eValue',
    );
  });

  it('should pass forcedHover', () => {
    // before
    const { container } = render(
      <ButtonIcon iconComponent={iconComponent} forcedHover />,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.buttonIcon)).toHaveClass(
      classNames[classNameButtonIcon].modificators.forcedHover,
    );
  });

  it('should pass iconComponent', () => {
    // before
    const { container } = render(
      <ButtonIcon
        classNameIcon={classNameIcon}
        iconComponent={iconComponent}
      />,
    );

    // result
    expect(container.querySelector(`.${classNameIcon}`)).toHaveClass(
      iconComponent.name,
    );
  });

  it('should pass onClick', () => {
    // before
    const { container } = render(
      <ButtonIcon iconComponent={iconComponent} onClick={mockCallBack} />,
    );

    // action
    fireEvent.click(getByE2EAttribute(container, E2EAttribute.buttonIcon));

    // result
    expect(mockCallBack.mock.calls.length).toBe(1);
  });

  it('should pass size', () => {
    // mock
    const sizes = enumToArray(ButtonSize) as Array<ButtonSize>;

    // before
    const { container } = render(
      <>
        {sizes.map((size) => (
          <ButtonIcon
            e2eValue={size}
            iconComponent={iconComponent}
            key={size}
            size={size}
          />
        ))}
      </>,
    );

    // result
    sizes.forEach((size) =>
      expect(
        getByE2EAttribute(container, E2EAttribute.buttonIcon, size),
      ).toHaveClass(classNames[classNameButtonIcon].modificators[size]),
    );
  });
});

describe('ButtonIcon snapshots', () => {
  it('should render Button', () => {
    // before
    const { asFragment } = render(<ButtonIcon iconComponent={iconComponent} />);

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
