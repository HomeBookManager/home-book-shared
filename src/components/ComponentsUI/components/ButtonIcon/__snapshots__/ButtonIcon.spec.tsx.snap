// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonIcon snapshots should render Button 1`] = `
<DocumentFragment>
  <button
    class="ButtonIcon ButtonIcon--medium"
    data-test-button-icon=""
    type="button"
  >
    <svg
      class="ButtonIcon__icon ButtonIcon__icon--medium IconUpOutlined IconUpOutlined--forced-hover"
      data-test-icon="icon-up-outlined"
      fill="none"
      height="16px"
      tabindex="-1"
      viewBox="64 64 896 896"
      width="16px"
    >
      <path
        class=""
        d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
        fill=""
      />
    </svg>
  </button>
</DocumentFragment>
`;
